llama_index/chat_engine/types.py

```diff
@@ -1,17 +1,25 @@
+import asyncio
 import logging
 import queue
 from abc import ABC, abstractmethod
 from dataclasses import dataclass, field
-from enum import Enum
-from typing import Generator, List, Optional
+from enum import Enum, auto
+from threading import Event
+from typing import AsyncGenerator, Generator, List, Optional, Union
 
 from llama_index.llms.base import ChatMessage, ChatResponseAsyncGen, ChatResponseGen
 from llama_index.memory import BaseMemory
-from llama_index.tools import ToolOutput
 from llama_index.response.schema import Response, StreamingResponse
 from llama_index.schema import NodeWithScore
+from llama_index.tools import ToolOutput
 
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.WARNING)
+
+
+def is_function(message: ChatMessage) -> bool:
+    """Utility for ChatMessage responses from OpenAI models"""
+    return message.additional_kwargs.get("function_call", None) is not None
```

Comment:
very nit: 
```
"function_call" in message.additional_kwargs
```

Reply:
Good call, I just shifted this from the original implementation into a standalone function to clean up the method body without really thinking to simplify it. Fixed now.